### ISSUE TITLE
Build failure when including a plugin build in settings & applying a settings plugin

### ISSUE THREAD

**Issue by 3flex** [gradle/gradle#16532]
Build failure when using this config, and the included build applies a plugin that calls `getRootProject`:
```
pluginManagement {
    includeBuild("greeting-plugin")
}

plugins {
    id("com.gradle.enterprise").version("3.6")
}
```

### Current Behavior or error
```
* What went wrong:
An exception occurred applying plugin request [id: 'org.jetbrains.kotlin.multiplatform']
> Failed to apply plugin 'org.jetbrains.kotlin.multiplatform'.
   > The root project is not yet available for build.
```
The error is due to applying `org.jetbrains.kotlin.multiplatform` plugin in the included Gradle plugin's build script. The stacktrace points to use of `getRootProject` on [this line](https://github.com/JetBrains/kotlin/blob/v1.4.31/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/KotlinProjectStructureMetadata.kt#L299) in the Kotlin multiplatform plugin.

However, this only happens when both conditions are true:
1. Including the plugin build using `includePlugin` in `pluginManagement` block
2. Applying a settings plugin. I used the Gradle Enterprise Plugin for the example, but any plugin could be applied in the settings string (even ones that don't exist) to trigger the error condition.

I can either comment the `id()` line in `plugins` block OR move `includeBuild` out of the `pluginManagement` block to include the build in the "old" way and the build will then work.

### Steps to Reproduce
https://github.com/3flex/gradle-include-plugin

Make changes in settings.gradle.

#### Works:
```groovy
pluginManagement {
    includeBuild("greeting-plugin")
}

plugins {
    // id("com.gradle.enterprise").version("3.6")
}

include("my-greeting-app")
```
#### Works:
```groovy
plugins {
    id("com.gradle.enterprise").version("3.6")
}

includeBuild("greeting-plugin")

include("my-greeting-app")
```
#### Fails:
```groovy
pluginManagement {
    includeBuild("greeting-plugin")
}

plugins {
    id("com.gradle.enterprise").version("3.6")
}

include("my-greeting-app")
```
### Your Environment
7.0-milestone-3
Windows 10 20H2
Running Gradle directly in the terminal (not IDE)
Build scan URL: N/A, build scan not generated due to early failure.

**Comment by jjohannes** [comment]
@3flex thanks for reporting this.

You only see it when you apply a settings plugin, because only then the included builds are evaluated to search for the plugins.

I am not sure if we can make this work (we should certainly give a better error). Maybe this would need to be handled better in the Kotlin MPP plugin.

Are you blocked on this in any form? Can you explain your use case and why you need to use the MPP plugin to write Gradle plugins?

Or is this just something you discovered by chance. :)

**Comment by 3flex** [comment]
I may not have been clear, the build I'm including is a project plugin, not a settings plugin. I understood including the project plugin build in `pluginManagement` should work based on the milestone 3 release notes:

> This release introduces a new DSL construct in the settings file for including plugin builds. Build included like that can provide both project and settings plugins.

The build failure happens when I use that new DSL construct to include a project plugin build at the same time as I try to apply any other settings plugin (it doesn't even have to be a valid plugin, `id("random-string")` will cause the originally reported error).

I don't directly need multiplatform Kotlin to develop a Gradle plugin, but one of the included project plugin build's dependencies applies the MPP plugin which calls getRootProject so it bubbles back up.

**Comment by jjohannes** [comment]
Yes I understand that this is not about `settings` plugins. The problem is, how the mechanism is implemented today, as soon as you apply any settings plugin, Gradle will search all the builds included in `pluginManagement {}`  for that plugin. And to do that, it needs to configure these builds. Because Gradle does not know what's inside all of these builds, it configures them no matter if they contain a settings plugin or not.

This step is performed later if you do not apply any settings plugins. That's why you only the the error if you apply **any** settings plugin.

> I don't directly need multiplatform Kotlin to develop a Gradle plugin, but one of the included project plugin build's dependencies applies the MPP plugin which calls getRootProject so it bubbles back up.

Thanks for the explanation. I see how that can happen.

**Comment by TimvdLippe** [comment]
Okay, I have been able to figure out a workaround: remove the `dependencySubstitution` block from the `includeBuild`. I am not sure why, but if you include that, the IntelliJ Gradle import breaks. Removing it (and setting the appropriate `rootProject.group` in the included build) works as intended.
